File: sentry-replays-meticulous/sentry_client.py
#!/usr/bin/env python3
"""
Sentry Replays API Client

Fetches session replay data from Sentry's API for analysis and replay testing.
"""

import os
import json
import requests
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta


class SentryReplayClient:
    """Client for interacting with Sentry's Replays API."""

    def __init__(self, auth_token: Optional[str] = None, organization_slug: Optional[str] = None):
        """
        Initialize Sentry client.

        Args:
            auth_token: Sentry auth token (or set SENTRY_AUTH_TOKEN env var)
            organization_slug: Organization slug (or set SENTRY_ORG_SLUG env var)
        """
        self.auth_token = auth_token or os.getenv('SENTRY_AUTH_TOKEN')
        self.org_slug = organization_slug or os.getenv('SENTRY_ORG_SLUG')
        self.base_url = 'https://sentry.io/api/0'

        if not self.auth_token:
            raise ValueError("auth_token required (or set SENTRY_AUTH_TOKEN)")
        if not self.org_slug:
            raise ValueError("organization_slug required (or set SENTRY_ORG_SLUG)")

    def _get_headers(self) -> Dict[str, str]:
        """Get HTTP headers for API requests."""
        return {
            'Authorization': f'Bearer {self.auth_token}',
            'Content-Type': 'application/json'
        }

    def list_replays(
        self,
        project: Optional[List[int]] = None,
        environment: Optional[str] = None,
        stats_period: str = '7d',
        start: Optional[str] = None,
        end: Optional[str] = None,
        query: Optional[str] = None,
        sort: Optional[str] = None,
        per_page: int = 50,
        cursor: Optional[str] = None,
        fields: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """
        List replays for an organization.

        Args:
            project: List of project IDs to filter by
            environment: Environment name to filter by
            stats_period: Time range (e.g., '1d', '7d', '30d') - used if start/end not provided
            start: Start datetime (ISO8601 or epoch seconds)
            end: End datetime (ISO8601 or epoch seconds)
            query: Query string for filtering (e.g., 'error_count:>0')
            sort: Field to sort by
            per_page: Results per page (max 100)
            cursor: Pagination cursor
            fields: Specific fields to return

        Returns:
            Dict with 'data' (list of replays) and pagination info
        """
        url = f'{self.base_url}/organizations/{self.org_slug}/replays/'

        params = {}

        # Time range
        if start and end:
            params['start'] = start
            params['end'] = end
        else:
            params['statsPeriod'] = stats_period

        # Filters
        if project:
            params['project'] = project
        if environment:
            params['environment'] = environment
        if query:
            params['query'] = query
        if sort:
            params['sort'] = sort

        # Pagination and fields
        params['per_page'] = per_page
        if cursor:
            params['cursor'] = cursor
        if fields:
            params['field'] = fields

        response = requests.get(url, headers=self._get_headers(), params=params)
        response.raise_for_status()

        # Parse Link header for pagination
        link_header = response.headers.get('Link', '')
        pagination = self._parse_link_header(link_header)

        return {
            'data': response.json(),
            'pagination': pagination
        }

    def get_replay_details(self, replay_id: str) -> Dict[str, Any]:
        """
        Get detailed information about a specific replay.

        Args:
            replay_id: The replay ID

        Returns:
            Replay details
        """
        url = f'{self.base_url}/organizations/{self.org_slug}/replays/{replay_id}/'
        response = requests.get(url, headers=self._get_headers())
        response.raise_for_status()
        return response.json()

    def get_replay_events(self, replay_id: str) -> List[Dict[str, Any]]:
        """
        Get the event stream for a replay (user actions, network requests, etc.).

        Note: This endpoint may vary based on Sentry version. Check API docs.

        Args:
            replay_id: The replay ID

        Returns:
            List of replay events
        """
        # This is a placeholder - actual endpoint may differ
        url = f'{self.base_url}/organizations/{self.org_slug}/replays/{replay_id}/events/'

        try:
            response = requests.get(url, headers=self._get_headers())
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 404:
                print(f"Events endpoint not available. Check Sentry API docs for replay event access.")
                return []
            raise

    def _parse_link_header(self, link_header: str) -> Dict[str, Optional[str]]:
        """
        Parse Link header for pagination cursors.

        Args:
            link_header: Raw Link header value

        Returns:
            Dict with 'next' and 'previous' cursors
        """
        pagination = {'next': None, 'previous': None}

        if not link_header:
            return pagination

        # Link header format: <url>; rel="next", <url>; rel="previous"
        links = link_header.split(',')

        for link in links:
            parts = link.split(';')
            if len(parts) != 2:
                continue

            url = parts[0].strip().strip('<>')
            rel = parts[1].strip()

            if 'rel="next"' in rel:
                # Extract cursor from URL
                if 'cursor=' in url:
                    cursor = url.split('cursor=')[1].split('&')[0]
                    pagination['next'] = cursor
            elif 'rel="previous"' in rel:
                if 'cursor=' in url:
                    cursor = url.split('cursor=')[1].split('&')[0]
                    pagination['previous'] = cursor

        return pagination

    def fetch_all_replays(
        self,
        max_pages: int = 10,
        **kwargs
    ) -> List[Dict[str, Any]]:
        """
        Fetch all replays across multiple pages.

        Args:
            max_pages: Maximum number of pages to fetch
            **kwargs: Arguments passed to list_replays()

        Returns:
            List of all replay objects
        """
        all_replays = []
        cursor = None

        for page in range(max_pages):
            result = self.list_replays(cursor=cursor, **kwargs)
            replays = result['data']

            if not replays:
                break

            all_replays.extend(replays)

            # Check for next page
            next_cursor = result['pagination'].get('next')
            if not next_cursor:
                break

            cursor = next_cursor
            print(f"Fetched page {page + 1}, total replays: {len(all_replays)}")

        return all_replays


def main():
    """Example usage of SentryReplayClient."""

    # Initialize client (reads from env vars)
    try:
        client = SentryReplayClient()
    except ValueError as e:
        print(f"Error: {e}")
        print("\nSet environment variables:")
        print("  export SENTRY_AUTH_TOKEN='your-token'")
        print("  export SENTRY_ORG_SLUG='your-org-slug'")
        return

    print(f"Fetching replays for organization: {client.org_slug}")

    # Fetch replays from last 7 days
    result = client.list_replays(
        stats_period='7d',
        per_page=10,
        fields=['id', 'project_id', 'user', 'browser', 'os', 'duration', 'error_count']
    )

    replays = result['data']
    print(f"\nFound {len(replays)} replays")

    # Display summary
    for replay in replays:
        print(f"\nReplay ID: {replay.get('id')}")
        print(f"  Project: {replay.get('project_id')}")
        print(f"  User: {replay.get('user', {})}")
        print(f"  Browser: {replay.get('browser', {}).get('name')} {replay.get('browser', {}).get('version')}")
        print(f"  Duration: {replay.get('duration')}ms")
        print(f"  Errors: {replay.get('error_count')}")

    # Get detailed info for first replay
    if replays:
        first_replay_id = replays[0].get('id')
        if first_replay_id:
            print(f"\n\nFetching details for replay: {first_replay_id}")
            details = client.get_replay_details(first_replay_id)
            print(json.dumps(details, indent=2))


if __name__ == '__main__':
    main()

File: sentry-replays-meticulous/src/sentry-client.ts
/**
 * Sentry Replays API Client
 *
 * Fetches session replay data from Sentry's API for analysis and replay testing.
 */

import axios, { AxiosInstance } from 'axios';

interface SentryReplay {
  id: string;
  project_id: number;
  duration: number;
  finished_at: string;
  started_at: string;
  user?: {
    email?: string;
    id?: string;
    username?: string;
  };
  browser?: {
    name: string;
    version: string;
  };
  os?: {
    name: string;
    version: string;
  };
  device?: {
    name: string;
    brand: string;
    model: string;
  };
  error_count: number;
  urls?: string[];
  [key: string]: any;
}

interface PaginationInfo {
  next: string | null;
  previous: string | null;
}

interface ListReplaysResult {
  data: SentryReplay[];
  pagination: PaginationInfo;
}

interface ListReplaysOptions {
  project?: number[];
  environment?: string;
  statsPeriod?: string;
  start?: string;
  end?: string;
  query?: string;
  sort?: string;
  perPage?: number;
  cursor?: string;
  fields?: string[];
}

interface FetchAllReplaysOptions extends Omit<ListReplaysOptions, 'cursor'> {
  maxPages?: number;
}

export class SentryReplayClient {
  private authToken: string;
  private orgSlug: string;
  private baseUrl: string = 'https://sentry.io/api/0';
  private client: AxiosInstance;

  constructor(authToken?: string, organizationSlug?: string) {
    this.authToken = authToken || process.env.SENTRY_AUTH_TOKEN || '';
    this.orgSlug = organizationSlug || process.env.SENTRY_ORG_SLUG || '';

    if (!this.authToken) {
      throw new Error('authToken required (or set SENTRY_AUTH_TOKEN env var)');
    }
    if (!this.orgSlug) {
      throw new Error('organizationSlug required (or set SENTRY_ORG_SLUG env var)');
    }

    this.client = axios.create({
      baseURL: this.baseUrl,
      headers: {
        'Authorization': `Bearer ${this.authToken}`,
        'Content-Type': 'application/json',
      },
    });
  }

  /**
   * List replays for an organization.
   */
  async listReplays(options: ListReplaysOptions = {}): Promise<ListReplaysResult> {
    const {
      project,
      environment,
      statsPeriod = '7d',
      start,
      end,
      query,
      sort,
      perPage = 50,
      cursor,
      fields,
    } = options;

    const params: Record<string, any> = {};

    // Time range
    if (start && end) {
      params.start = start;
      params.end = end;
    } else {
      params.statsPeriod = statsPeriod;
    }

    // Filters
    if (project) params.project = project;
    if (environment) params.environment = environment;
    if (query) params.query = query;
    if (sort) params.sort = sort;

    // Pagination and fields
    params.per_page = perPage;
    if (cursor) params.cursor = cursor;
    if (fields) params.field = fields;

    const response = await this.client.get(
      `/organizations/${this.orgSlug}/replays/`,
      { params }
    );

    // Parse Link header for pagination
    const linkHeader = response.headers.link || '';
    const pagination = this.parseLinkHeader(linkHeader);

    return {
      data: response.data,
      pagination,
    };
  }

  /**
   * Get detailed information about a specific replay.
   */
  async getReplayDetails(replayId: string): Promise<SentryReplay> {
    const response = await this.client.get(
      `/organizations/${this.orgSlug}/replays/${replayId}/`
    );
    return response.data;
  }

  /**
   * Get the event stream for a replay (user actions, network requests, etc.).
   *
   * Note: This endpoint may vary based on Sentry version.
   */
  async getReplayEvents(replayId: string): Promise<any[]> {
    try {
      const response = await this.client.get(
        `/organizations/${this.orgSlug}/replays/${replayId}/events/`
      );
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        console.warn('Events endpoint not available. Check Sentry API docs for replay event access.');
        return [];
      }
      throw error;
    }
  }

  /**
   * Parse Link header for pagination cursors.
   */
  private parseLinkHeader(linkHeader: string): PaginationInfo {
    const pagination: PaginationInfo = {
      next: null,
      previous: null,
    };

    if (!linkHeader) {
      return pagination;
    }

    // Link header format: <url>; rel="next", <url>; rel="previous"
    const links = linkHeader.split(',');

    for (const link of links) {
      const parts = link.split(';');
      if (parts.length !== 2) continue;

      const url = parts[0].trim().replace(/^<|>$/g, '');
      const rel = parts[1].trim();

      if (rel.includes('rel="next"') && url.includes('cursor=')) {
        const cursorMatch = url.match(/cursor=([^&]+)/);
        if (cursorMatch) pagination.next = cursorMatch[1];
      } else if (rel.includes('rel="previous"') && url.includes('cursor=')) {
        const cursorMatch = url.match(/cursor=([^&]+)/);
        if (cursorMatch) pagination.previous = cursorMatch[1];
      }
    }

    return pagination;
  }

  /**
   * Fetch all replays across multiple pages.
   */
  async fetchAllReplays(options: FetchAllReplaysOptions = {}): Promise<SentryReplay[]> {
    const { maxPages = 10, ...listOptions } = options;
    const allReplays: SentryReplay[] = [];
    let cursor: string | null = null;

    for (let page = 0; page < maxPages; page++) {
      const result = await this.listReplays({ ...listOptions, cursor: cursor || undefined });
      const replays = result.data;

      if (replays.length === 0) break;

      allReplays.push(...replays);

      // Check for next page
      const nextCursor = result.pagination.next;
      if (!nextCursor) break;

      cursor = nextCursor;
      console.log(`Fetched page ${page + 1}, total replays: ${allReplays.length}`);
    }

    return allReplays;
  }
}

/**
 * Example usage of SentryReplayClient
 */
async function main() {
  try {
    const client = new SentryReplayClient();

    console.log(`Fetching replays for organization: ${(client as any).orgSlug}`);

    // Fetch replays from last 7 days
    const result = await client.listReplays({
      statsPeriod: '7d',
      perPage: 10,
      fields: ['id', 'project_id', 'user', 'browser', 'os', 'duration', 'error_count'],
    });

    const replays = result.data;
    console.log(`\nFound ${replays.length} replays`);

    // Display summary
    for (const replay of replays) {
      console.log(`\nReplay ID: ${replay.id}`);
      console.log(`  Project: ${replay.project_id}`);
      console.log(`  User: ${JSON.stringify(replay.user)}`);
      console.log(`  Browser: ${replay.browser?.name} ${replay.browser?.version}`);
      console.log(`  Duration: ${replay.duration}ms`);
      console.log(`  Errors: ${replay.error_count}`);
    }

    // Get detailed info for first replay
    if (replays.length > 0 && replays[0].id) {
      const firstReplayId = replays[0].id;
      console.log(`\n\nFetching details for replay: ${firstReplayId}`);
      const details = await client.getReplayDetails(firstReplayId);
      console.log(JSON.stringify(details, null, 2));
    }
  } catch (error: any) {
    if (error.message.includes('required')) {
      console.error(`Error: ${error.message}\n`);
      console.log('Set environment variables:');
      console.log("  export SENTRY_AUTH_TOKEN='your-token'");
      console.log("  export SENTRY_ORG_SLUG='your-org-slug'");
    } else {
      throw error;
    }
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}
